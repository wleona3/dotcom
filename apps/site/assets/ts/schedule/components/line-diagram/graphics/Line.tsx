import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { LineDiagramStop } from "../../__schedule";
import {
  areOnDifferentBranchLines,
  isMergeStop,
  isBranchTerminusStop
} from "../line-diagram-helpers";
import {
  StopCoord,
  CoordState,
  BRANCH_LINE_WIDTH,
  BASE_LINE_WIDTH,
  BRANCH_SPACING
} from "./graphic-helpers";

interface PathGraphicsProps {
  from: LineDiagramStop;
  to: LineDiagramStop;
}

const showCRLine = (from: LineDiagramStop, to: LineDiagramStop): boolean => {
  const OLNoCR = ["place-haecl", "place-state","place-dwnxg","place-chncl","place-tumnl"]
  return !(OLNoCR.includes(from.route_stop.id) || OLNoCR.includes(to.route_stop.id));
}

const showShuttleLine = (from: LineDiagramStop, to: LineDiagramStop): boolean => {
  const OLNoShuttle = ["place-state","place-dwnxg","place-chncl","place-tumnl"]
  return !(OLNoShuttle.includes(from.route_stop.id) || OLNoShuttle.includes(to.route_stop.id));
}
const Line = ({ from, to }: PathGraphicsProps): ReactElement | null => {
  const fromCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[from.route_stop.id]
  );
  const toCoords: StopCoord | null = useSelector(
    (state: CoordState) => state[to.route_stop.id]
  );
  if (!fromCoords || !toCoords) return null;
  if (isMergeStop(to) && from.stop_data.length === to.stop_data.length) {
    return null;
  }

  const [x1, y1] = fromCoords;
  const y2 = toCoords[1];

  if (isMergeStop(from) && !areOnDifferentBranchLines(from, to)) {
    // just draw one line.
    const lineProps: React.SVGProps<SVGLineElement> = {
      key: `${from.route_stop.id}-${to.route_stop.id}`
    };
    lineProps.strokeWidth = `${BASE_LINE_WIDTH}px`;
    lineProps.x1 = `${x1}px`;
    lineProps.x2 = lineProps.x1;
    lineProps.y1 = `${y1}px`;
    lineProps.y2 = `${y2}px`;
    // if (
    //   // from.stop_data.some(sd => sd["has_disruption?"]) &&
    //   // to.stop_data.some(sd => sd["has_disruption?"])
    //   true
    // ) {
    //   lineProps.stroke = "url(#diagonalHatch)";
    // }
    lineProps.stroke = "url(#diagonalHatch)";
    return <line className="line-diagram-svg__line" {...lineProps} />;
  }

  // otherwise let's just show all the lines...
  return (
    <>
      {from.stop_data.map((branchLine, branchIndex) => {
        const lineProps: React.SVGProps<SVGLineElement> = {
          key: `${from.route_stop.id}-${to.route_stop.id}-${branchLine.type}-${branchIndex}`
        };
        lineProps.stroke = "url(#diagonalHatch)";
        lineProps.strokeWidth = `${
          branchIndex >= 1 &&
          !(isBranchTerminusStop(from) || isBranchTerminusStop(to)) &&
          (branchLine.type === "stop" ||
            to.stop_data[branchIndex].type === "stop")
            ? BRANCH_LINE_WIDTH
            : BASE_LINE_WIDTH
        }px`;
        lineProps.x1 = `${BRANCH_SPACING * branchIndex +
          BASE_LINE_WIDTH +
          1}px`;
        lineProps.x2 = lineProps.x1;
        lineProps.y1 = `${y1}px`;
        lineProps.y2 = `${y2}px`;
        // if (branchLine["has_disruption?"]) {
        //   lineProps.stroke = "url(#diagonalHatch)";
        // }

        const linePropsCR = { ...lineProps };
        linePropsCR.x1 = `${BRANCH_SPACING * (branchIndex + 1) +
          BASE_LINE_WIDTH +
          1}px`;
        linePropsCR.x2 = linePropsCR.x1;
        linePropsCR.stroke = "purple";
        return <>
          {showShuttleLine(from, to) ? <line key={lineProps.key} className="line-diagram-svg__line" {...lineProps} /> : null}
          {showCRLine(from, to) ? <line key={`${from.route_stop.id}-${to.route_stop.id}-${branchLine.type}-${branchIndex + 1}--cr`} className="line-diagram-svg__line u-text--commuter-rail" {...linePropsCR} />: null}
        </>;
      })}
    </>
  );
};

export default Line;
