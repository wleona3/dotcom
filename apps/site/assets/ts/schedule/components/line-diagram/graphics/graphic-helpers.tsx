import React from "react";

export const CIRC_RADIUS = 4;
export const CIRC_DIAMETER = CIRC_RADIUS * 2;
export const BASE_LINE_WIDTH = CIRC_DIAMETER + 2;
export const BRANCH_LINE_WIDTH = BASE_LINE_WIDTH + 6;
export const BRANCH_SPACING = 16;
export const MERGE_RADIUS = 6;

export type StopCoord = [number, number];
export interface CoordState {
  [key: string]: StopCoord | null;
}

export const DiagonalHatchPattern = (
  id?: string | undefined
): React.SVGProps<SVGPatternElement> => (
  <pattern
    id={id ?? "diagonalHatch"}
    width="10"
    height="10"
    patternTransform="rotate(45 0 0)"
    patternUnits="userSpaceOnUse"
  >
    <line
      className="line-diagram-svg__line"
      y2={CIRC_DIAMETER}
      strokeWidth={`${BRANCH_LINE_WIDTH}px`}
    />
  </pattern>
);
