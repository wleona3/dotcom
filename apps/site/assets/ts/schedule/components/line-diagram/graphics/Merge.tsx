import { first, last } from "lodash";
import React, { ReactElement } from "react";
import { LineDiagramStop } from "../../__schedule";
import { useStopPositionCoordinates } from "../contexts/StopPositionContext";
import {
  getTreeDirection,
  isBranchTerminusStop,
  isMergeStop,
  lineDiagramIndexes,
  isStopOnMainLine
} from "../line-diagram-helpers";

import { StopCoord, BASE_LINE_WIDTH, MERGE_RADIUS } from "./graphic-helpers";
import Line from "./Line";

interface MergeGraphicsProps {
  lineDiagram: LineDiagramStop[];
}
const Merges = ({ lineDiagram }: MergeGraphicsProps): ReactElement | null => {
  const branchingOutward = getTreeDirection(lineDiagram) === "outward";
  const mergeIndices = lineDiagramIndexes(lineDiagram, isMergeStop);
  // walk through the diagram to find where all the merges go.
  const stopGaps = mergeIndices.map(i => {
    let from: LineDiagramStop;
    let to: LineDiagramStop;
    if (branchingOutward) {
      from = lineDiagram[i];
      [to] = lineDiagram.slice(i).filter(isStopOnMainLine);
    } else {
      to = lineDiagram[i];
      from = last(lineDiagram.slice(0, i - 1).filter(isStopOnMainLine))!;
    }

    return [from, to];
  });

  const allCoordinates = useStopPositionCoordinates();
  if (!allCoordinates || !allCoordinates[lineDiagram[0].route_stop.id])
    return null;

  const mergeBends = stopGaps.map(([from, to]) => {
    // if outward, start with from and construct array of tos
    // if inward, start with to and construct array of froms
    const mergeStop = branchingOutward ? from : to;
    const mergeStopIndex = lineDiagram.indexOf(mergeStop);
    const mergeStopCoords = allCoordinates[mergeStop.route_stop.id];
    if (!mergeStopCoords) return null;

    // usually there's just one terminus to draw the merge to... except for the GL at Kenmore.
    const mergeEnds = mergeStop.stop_data.slice(1).map(({ branch }) => {
      // along the branch might be subsequent stops or prior stops
      const downBranch = (branchingOutward
        ? lineDiagram.slice(mergeStopIndex)
        : lineDiagram.slice(0, mergeStopIndex)
      ).filter(stop => stop.route_stop.branch === branch);
      const nextStopOnBranch = branchingOutward
        ? first(downBranch)
        : last(downBranch);
      const terminusOnBranch = branchingOutward
        ? first(downBranch.filter(isBranchTerminusStop))
        : last(downBranch.filter(isBranchTerminusStop));
      return [nextStopOnBranch, terminusOnBranch];
    });

    return mergeEnds.map(([nextStop, terminus]) => {
      if (!nextStop && !terminus) return null;
      // try the next stop on the branch first. if the branch is collapsed the
      // coordinates will be null as the next stop is hidden, so use terminus
      const nextStopVisible = !!allCoordinates[nextStop!.route_stop.id];
      const nextStopCoords = nextStopVisible
        ? allCoordinates[nextStop!.route_stop.id]
        : allCoordinates[terminus!.route_stop.id];

      const [x, y] = nextStopCoords as StopCoord;
      const [mX, mY] = mergeStopCoords;
      const dx = (Math.abs(mX - x) - MERGE_RADIUS) * -1;
      const dy =
        (Math.abs(mY - y) - MERGE_RADIUS - (branchingOutward ? 60 : 20)) *
        (branchingOutward ? -1 : 1);
      const arc = branchingOutward
        ? `-${MERGE_RADIUS},-${MERGE_RADIUS} 0 0 0 -${MERGE_RADIUS},-${MERGE_RADIUS}`
        : `-${MERGE_RADIUS},${MERGE_RADIUS} 0 0 1 -${MERGE_RADIUS},${MERGE_RADIUS}`;
      const pathProps: React.SVGProps<SVGPathElement> = {
        key: `${mergeStop.route_stop.id}-${terminus!.route_stop.id}-merge`,
        strokeWidth: `${BASE_LINE_WIDTH}px`,
        d: `M${x},${y} v${dy} a${arc} h${dx}`
      };
      if (nextStop!.stop_data.some(sd => sd["has_disruption?"])) {
        pathProps.stroke = "url(#diagonalHatch)";
      }
      return <path {...pathProps} />;
    });
  });

  if (!stopGaps.length && !mergeBends.length) return null;

  return (
    <g className="line-diagram-svg__merge">
      {stopGaps.map(
        ([from, to]) =>
          from &&
          to && (
            <Line
              key={`${from.route_stop.id}-${to.route_stop.id}-line`}
              from={from}
              to={to}
            />
          )
      )}
      {mergeBends}
    </g>
  );
};

export default Merges;
