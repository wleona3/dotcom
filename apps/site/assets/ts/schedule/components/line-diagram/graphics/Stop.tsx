import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { LineDiagramStop } from "../../__schedule";
import { StopCoord, CoordState, CIRC_RADIUS, BRANCH_SPACING } from "./graphic-helpers";

const CRStops = ["place-ogmnl", "place-mlmnl", "place-north", "place-bbsta", "place-rugg", "place-forhl"]
const Stop = ({
  stop
}: {
  stop: LineDiagramStop;
}): ReactElement<SVGCircleElement> | null => {
  const coords: StopCoord | null = useSelector(
    (state: CoordState) => state[stop.route_stop.id]
  );
  if (!coords) return null;
  const [x, y] = coords;
  return (
    <>
      <circle
        key={stop.route_stop.id}
        className="line-diagram-svg__stop"
        r={`${CIRC_RADIUS}px`}
        cx={`${x}px`}
        cy={`${y}px`}
      />
      {CRStops.includes(stop.route_stop.id) ? (
        <circle
        key={`${stop.route_stop.id}-cr`}
        className="line-diagram-svg__stop"
        r={`${CIRC_RADIUS}px`}
        cx={`${x + BRANCH_SPACING}px`}
        cy={`${y}px`}
      />
      ) : null}
    </>
    
  );
};

export default Stop;
