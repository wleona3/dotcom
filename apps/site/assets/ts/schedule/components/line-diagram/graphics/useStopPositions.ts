import {
  createContext,
  useRef,
  useEffect,
  useCallback,
  useState,
  useLayoutEffect
} from "react";
import { LineDiagramStop } from "../../__schedule";
import { isMergeStop } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, BRANCH_SPACING, StopCoord } from "./graphic-helpers";

const xCoordForStop = (stop: LineDiagramStop): number => {
  if (isMergeStop(stop)) {
    return BASE_LINE_WIDTH + 1;
  }
  // usually just a function of stop_data length
  return BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
};

type RefMapSetter = (stopId: string) => (el: HTMLElement | null) => void;
type RefMapUpdator = () => void;
type CoordGet = (stopId: string) => StopCoord | null;
export const StopRefContext = createContext<
  [RefMapSetter, RefMapUpdator, CoordGet]
>([() => () => {}, () => {}, () => null]);

export default function useStopPositions(
  stops: LineDiagramStop[]
): [RefMapSetter, RefMapUpdator, CoordGet] {
  const stopRefsMap = useRef(new Map());
  const initialCoordState = useRef<{
    [stopId: string]: StopCoord | null;
  }>({});

  // only if stops change!
  useEffect(() => {
    stops.forEach(stop => {
      const stopId = stop.route_stop.id;
      stopRefsMap.current.set(stopId, null);
      initialCoordState.current[stopId] = null;
    });
  }, [initialCoordState, stops]);

  const [stopCoordState, setStopCoordState] = useState(
    initialCoordState.current
  );

  const updateAllStops = useCallback(
    () =>
      window.requestAnimationFrame((): void => {
        setStopCoordState(state => {
          const newState = { ...state };
          stops.forEach(stop => {
            const stopId = stop.route_stop.id;
            const el = stopRefsMap.current.get(stopId);
            if (el) {
              const { offsetTop, offsetHeight } = el;
              const y = offsetTop + offsetHeight / 2;
              const x = xCoordForStop(stop);
              if (x && y) {
                newState[stopId] = [x, y];
              } else {
                newState[stopId] = null;
              }
            } else {
              newState[stopId] = null;
            }
          });
          return newState;
        });
      }),
    [stops]
  );

  useLayoutEffect(() => {
    window.addEventListener("resize", updateAllStops);
    return () => window.removeEventListener("resize", updateAllStops);
  }, [updateAllStops]);

  const setupStopRef: RefMapSetter = stopId => el => {
    stopRefsMap.current.set(stopId, el);
  };

  const getCoord: CoordGet = stopId => stopCoordState[stopId];

  return [setupStopRef, updateAllStops, getCoord];
}
