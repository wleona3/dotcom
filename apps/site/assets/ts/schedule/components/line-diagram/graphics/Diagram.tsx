import { max } from "lodash";
import React, { ReactElement } from "react";
import { useSelector } from "react-redux";
import { LineDiagramStop } from "../../__schedule";
import {
  hasBranchLines,
  isBranchTerminusStop,
  areOnDifferentBranchLines,
  diagramWidth,
  isMergeStop
} from "../line-diagram-helpers";
import Line from "./Line";
import Merges from "./Merge";
import Stop from "./Stop";
import { routeToModeName } from "../../../../helpers/css";
import { LiveDataByStop } from "../__line-diagram";
import VehicleIcons from "../VehicleIcons";
import { getPageDirection } from "../../../store/schedule-store";
import { DirectionId } from "../../../../__v3api";
import { isAGreenLineRoute } from "../../../../models/route";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";

interface DiagramProps {
  lineDiagram: LineDiagramStop[];
  liveData: LiveDataByStop | undefined;
  overrideStyle?: "shuttle" | "commuter-rail";
  overridePlacement?: number;
}

const branchingDescription = (lineDiagram: LineDiagramStop[]): string => {
  const mergeStops = lineDiagram.filter(isMergeStop);
  const branchText = mergeStops.map(mergeStop => {
    const mergeStopName = mergeStop.route_stop.name;
    const branchNames = mergeStop.stop_data
      .filter(sd => sd.branch)
      .map(sd => sd.branch);
    return `Use ${mergeStopName} to transfer to trains on ${branchNames.join(
      " or "
    )} branches. `;
  });
  return branchText.join("");
};
const diagramDescription = (
  lineDiagram: LineDiagramStop[],
  direction: DirectionId
): string => {
  const text = `${lineDiagram.length} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = lineDiagram[0].route_stop.route!;
  if (destinations) {
    return `${text} going ${names[direction]} to ${destinations[direction]}`;
  }
  return text;
};

interface VehicleIconSetProps {
  stop: LineDiagramStop;
  liveData: LiveDataByStop | undefined;
}

const LiveVehicleIconSet = ({
  stop,
  liveData
}: VehicleIconSetProps): ReactElement<HTMLElement> | null => {
  const stopId = stop.route_stop.id;
  if (!liveData || !liveData[stopId]) return null;
  // Hide vehicles arriving to the origin from 'off the line'
  const vehicleData = stop.route_stop["is_beginning?"]
    ? liveData[stopId].vehicles.filter(vehicle => vehicle.status === "stopped")
    : liveData[stopId].vehicles;

  return (
    <VehicleIcons
      key={`${stopId}-vehicles`}
      stop={stop.route_stop}
      vehicles={vehicleData}
    />
  );
};

const Diagram = (props: DiagramProps): ReactElement<HTMLElement> | null => {
  const { lineDiagram, liveData, overrideStyle, overridePlacement } = props;
  const selectedDirection = useSelector(getPageDirection);
  const width = diagramWidth(
    max(lineDiagram.map(ld => ld.stop_data.length)) || 1
  );

  /* istanbul ignore next */
  const extraClassName =
    overrideStyle === "commuter-rail"
      ? "commuter-rail"
      : routeToModeName(lineDiagram[0].route_stop.route!);

  return (
    <>
      {lineDiagram.map(stop => (
        <LiveVehicleIconSet
          key={stop.route_stop.id}
          stop={stop}
          liveData={liveData}
        />
      ))}
      <svg
        xmlns="http://www.w3.org/2000/svg"
        role="img"
        aria-labelledby="diagram-title diagram-desc"
        className={`line-diagram-svg ${extraClassName}`}
        width={`${width + 4}px`}
        height="100%"
        style={{
          left:
            BASE_LINE_WIDTH / 2 +
            /* istanbul ignore next */ (overridePlacement ?? 0)
        }}
      >
        <title id="diagram-title">
          Line diagram for{" "}
          {isAGreenLineRoute(lineDiagram[0].route_stop.route!)
            ? "Green Line"
            : lineDiagram[0].route_stop.route!.name}
        </title>
        <desc id="diagram-desc">
          {diagramDescription(lineDiagram, selectedDirection)}
          {hasBranchLines(lineDiagram) && branchingDescription(lineDiagram)}
        </desc>

        <defs>{DiagonalHatchPattern(overrideStyle)}</defs>
        {/* If there are multiple branches, draw the lines and curves for those */
        hasBranchLines(lineDiagram) && <Merges lineDiagram={lineDiagram} />}

        {/* Draw lines between stops */
        lineDiagram.map((current, idx, self) => {
          if (self[idx + 1]) {
            if (
              isBranchTerminusStop(current) &&
              areOnDifferentBranchLines(current, self[idx + 1])
            )
              return null; // don't draw a line from a branch end to the main line

            return (
              <Line
                key={`${current.route_stop.id}-${self[idx + 1].route_stop.id}`}
                from={current}
                to={self[idx + 1]}
                shuttle={overrideStyle === "shuttle"}
              />
            );
          }
          return null; // this is the end of the line, nothing to draw a line to
        })}

        {/* Draw circles for each stop */
        lineDiagram.map(stop => (
          <Stop
            key={stop.route_stop.id}
            stop={stop}
            shuttle={overrideStyle === "shuttle"}
          />
        ))}
      </svg>
    </>
  );
};

export default Diagram;
