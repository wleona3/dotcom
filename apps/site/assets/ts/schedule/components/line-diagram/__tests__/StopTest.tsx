import React from "react";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import Stop from "../graphics/Stop";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";
import { CIRC_RADIUS } from "../graphics/graphic-helpers";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];

test("Stop component renders and matches snapshot", () => {
  const { asFragment } = mockedRenderWithStopPositionContext(
    <svg>
      <Stop stop={lineDiagram[0]} />
    </svg>,
    [lineDiagram[0]]
  );

  expect(asFragment()).toMatchSnapshot();
});

test("Stop component shows an SVG circle for the stop", () => {
  const { container } = mockedRenderWithStopPositionContext(
    <svg>
      <Stop stop={lineDiagram[0]} />
    </svg>,
    [lineDiagram[0]]
  );

  const circle = container.getElementsByTagName("circle")[0];
  expect(circle).toBeDefined();
  expect(circle.classList).toContain("line-diagram-svg__stop");
  expect(circle.getAttribute("r")).toEqual(`${CIRC_RADIUS}px`);
  // mocked values
  expect(circle.getAttribute("cx")).toEqual("13px");
  expect(circle.getAttribute("cy")).toEqual("7px");
});
