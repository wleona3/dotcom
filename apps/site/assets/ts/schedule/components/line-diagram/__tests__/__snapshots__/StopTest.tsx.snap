// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stop component renders and matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <circle
      class="line-diagram-svg__stop"
      cx="13px"
      cy="7px"
      data-shuttle="false"
      r="4px"
    />
  </svg>
</DocumentFragment>
`;
